pkg/controller/registry/registry_client.go

```diff
@@ -0,0 +1,105 @@
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -o resolver/fakes/fake_registry_client_interface.go . RegistryClientInterface
+package registry
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	registryapi "github.com/operator-framework/operator-registry/pkg/api"
+	"github.com/operator-framework/operator-registry/pkg/client"
+	opregistry "github.com/operator-framework/operator-registry/pkg/registry"
+)
+
+type ChannelEntryStream interface {
+	Recv() (*registryapi.ChannelEntry, error)
+}
+
+type RegistryClientInterface interface {
```

Comment:
Let's compose the interface and types to avoid juggling two distinct clients (it looks like you were close already):

```go
import (
        "context"
	"fmt"
	"io"

        "google.golang.org/grpc"

	registryapi "github.com/operator-framework/operator-registry/pkg/api"
	"github.com/operator-framework/operator-registry/pkg/client"
	opregistry "github.com/operator-framework/operator-registry/pkg/registry"
)

type Interface interface { // Note: I hate this naming convention
    client.Interface // Include the method set of this other interface

    FindBundleThatProvides(ctx context.Context, group, version, kind, pkgName string) (*registryapi.Bundle, error)
    GetLatestChannelEntriesThatProvide(ctx context.Context, group, version, kind string) (*ChannelEntryIterator, error)
}

type Client struct {
    client.Interface // Embed an instance of the imported interface to satisfy its method set
}

func NewClientFromConn(conn *grpc.ClientConn) *Client {
    return &Client{
        Interface: client.NewClientFromConn(conn),
    }
}

var _ Interface = &Client{} // True
var _ client.Interface = &Client{} // Also true

// GetLatestChannelEntriesThatProvide uses registry client to get a list of
// latest channel entries that provide the requested API (via an iterator)
func (rc *Client) GetLatestChannelEntriesThatProvide(ctx context.Context, group, version, kind string) (*ChannelEntryIterator, error) {
	stream, err := rc.Registry.GetLatestChannelEntriesThatProvide(ctx, &registryapi.GetLatestProvidersRequest{Group: group, Version: version, Kind: kind})
	if err != nil {
		return nil, err
	}
	return NewChannelEntryIterator(stream), nil
}

// FindBundleThatProvides returns a bundle that provides the request API and
// doesn't belong to the provided package
func (rc *OLMRegistryClient) FindBundleThatProvides(ctx context.Context, group, version, kind, pkgName string) (*registryapi.Bundle, error) {
	it, err := rc.GetLatestChannelEntriesThatProvide(ctx, group, version, kind)
	if err != nil {
		return nil, err
	}
	entry := rc.filterChannelEntries(it, pkgName)
	if entry == nil {
		return nil, fmt.Errorf("Unable to find a channel entry which doesn't belong to package %s", pkgName)
	}
	bundle, err := rc.Registry.GetBundle(ctx, &registryapi.GetBundleRequest{PkgName: entry.PackageName, ChannelName: entry.ChannelName, CsvName: entry.BundleName})
	if err != nil {
		return nil, err
	}
	return bundle, nil
}
```